### ISSUE TITLE
Admin.GetMetadata: relationship between topic and allTopics parameters

### ISSUE THREAD

**Issue by hqthao** [confluentinc/confluent-kafka-go#374]
Description
===========

As the documentation of the function `GetMetaData`:

```
    // GetMetadata queries broker for cluster and topic metadata.
    // If topic is non-nil only information about that topic is returned, else if
    // allTopics is false only information about locally used topics is returned,
    // else information about all topics is returned.
    // GetMetadata is equivalent to listTopics, describeTopics and describeCluster in the Java API.
    func (a *AdminClient) GetMetadata(topic *string, allTopics bool, timeoutMs int) (*Metadata, error) {
    	return getMetadata(a, topic, allTopics, timeoutMs)
    }
```

So I guess if the topic is non-nil, the result always returns only the specific topic information. However, when I try to test, I see a very strange result. 

How to reproduce
================
Let say I I already have a Kafka topic. Then I call the method GetMetadata as the below code:
```
topic := "sample_topic"
admin.GetMetadata(nil, false, timeout)
admin.GetMetadata(nil, true, timeout)
admin.GetMetadata(&topic, false, timeout)
admin.GetMetadata(&topic, true, timeout)
```
Only `admin.GetMetadata(&topic, false, timeout)` return the specific topic information. All other cases return all topics information.

Then I try to test on the non-exist topic. Only `admin.GetMetadata(&topic, false, timeout)
` yields the error code.

```
data, err := admin.admin.GetMetadata(&topic, false, admin.timeOutMs())
fmt.Println(data.Topics[topic]) //   Broker: Unknown topic or partition
```

Then I reorder test cases, the result is also different.

Checklist
=========
Please provide the following information:

 - [ ] confluent-kafka-go and librdkafka version : v1.1.0
 - [ ] Apache Kafka broker version: 2.1
 - [ ] Operating system: MacOS

**Comment by edenhill** [comment]
Yes this API is a bit weird due to historical reasons, it will be changed at some point to actually make sense.

It has three different query modes:
 * topic is irrelevant, allTopics=true -> return information for all topics in cluster 
 * topic, allTopics=false -> return information for just topic
 * none, allTopics=false -> return information for all topics that have been used by the client. This is mostly relevant for producers that produce to multiple topics to refresh the metadata for the topics that it uses and nothing else.

From an application point of view you would either do topic,allTopics=false to query a single topic, or nil,allTopics=true to query all topics, the third mode is usually not relevant.

**Comment by hqt** [comment]
Thanks for the clear explanation. As I checked, I see that the admin client only queries on its connected node to get the information. Sometimes there is a slight delay between the controller to other brokers, hence delay for the admin client queries some metadata.

In that case, I might have an error when creating the topic, then get back metadata immediately. This situation rarely happens in real-life but happens on the unit test. I think this situation can be fixed by knowing the controller ID then creating the client which connects directly to this node.

Can we do that with `Admin.GetMetadata` function, or any functions?

**Comment by hqthao** [comment]
Thanks for the clear explanation. As I checked, I see that the admin client only queries on its connected node to get the information. Sometimes there is a slight delay between the controller to other brokers, hence delay for the admin client queries some metadata.

In that case, I might have an error when creating the topic, then get back metadata immediately. This situation rarely happens in real-life but happens on the unit test. I think this situation can be fixed by knowing the controller ID then creating the client which connects directly to this node.

Can we do that with Admin.GetMetadata function, or any functions?

**Comment by edenhill** [comment]
The CreateTopics API will only wait for the controller to acknowledge the topic creation, it may take some time for the topic to propagate to the rest of the cluster and during this time the topic may not be returned in metadata from other brokers.

There is currently no API in the Go client to retrieve the controller (rd_kafka_controllerid())

**Comment by hqt** [comment]
Thank you for your support. Your contribution is really awesome. I hope I will contribute to the project someday.
